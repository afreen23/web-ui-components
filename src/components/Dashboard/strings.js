export const ALL = 'All';
export const CRITICAL = 'Critical';
export const WARNING = 'Warning';
