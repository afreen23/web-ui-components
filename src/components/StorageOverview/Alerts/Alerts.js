import React from 'react';
import PropTypes from 'prop-types';
import { Row, Col } from 'patternfly-react';

import {
  DashboardCard,
  DashboardCardBody,
  DashboardCardHeader,
  DashboardCardTitle,
} from '../../Dashboard/DashboardCard';
import { DashboardCardActionsBody } from '../../Dashboard/DashboardCard/DashboardCardActionsBody';
import { Dropdown } from '../../Form/Dropdown';

import { StorageOverviewContext } from '../StorageOverviewContext';
import { AlertsBody } from '../../Dashboard/Alert/AlertsBody';
import { filterAlerts } from './utils';
import { AlertItem } from '../../Dashboard/Alert/AlertItem';
import { getAlertSeverity } from '../../../selectors/prometheus/alerts';
import { ALL, WARNING, CRITICAL } from '../../Dashboard/strings';

const alertsFilterOptions = [ALL, CRITICAL, WARNING];

export class Alerts extends React.PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      type: ALL,
    };
  }

  filterBy = (allAlerts, filter) => allAlerts.filter(alert => getAlertSeverity(alert) === filter);

  getAlerts = allAlerts => {
    switch (this.state.type) {
      case ALL:
        return allAlerts;
      case WARNING:
        return this.filterBy(allAlerts, 'warning');
      case CRITICAL:
        return this.filterBy(allAlerts, 'critical');
      default:
        return [];
    }
  };

  render() {
    const { alertsResponse, className } = this.props;

    if (!Array.isArray(alertsResponse)) {
      return null;
    }
    const allAlerts = filterAlerts(alertsResponse);

    if (allAlerts.length === 0) {
      return null;
    }
    const alerts = this.getAlerts(allAlerts);
    const isLoading = !alerts;
    return (
      <DashboardCard className={className}>
        <DashboardCardHeader>
          <Row>
            <Col lg={9} md={9} sm={9} xs={9}>
              <DashboardCardTitle>Alerts</DashboardCardTitle>
            </Col>
            <Col lg={3} md={3} sm={3} xs={3}>
              <DashboardCardActionsBody>
                <Dropdown
                  id="alert-type"
                  value={this.state.type}
                  choices={alertsFilterOptions}
                  onChange={newVal => this.setState({ type: newVal })}
                  disabled={isLoading}
                />
              </DashboardCardActionsBody>
            </Col>
          </Row>
        </DashboardCardHeader>
        <DashboardCardBody>
          <AlertsBody>
            {alerts.map((alert, index) => (
              <AlertItem key={`alert-${index}`} alert={alert} />
            ))}
          </AlertsBody>
        </DashboardCardBody>
      </DashboardCard>
    );
  }
}

Alerts.propTypes = {
  alertsResponse: PropTypes.oneOfType([PropTypes.array, PropTypes.object]),
  className: PropTypes.string,
};

Alerts.defaultProps = {
  alertsResponse: null,
  className: null,
};

export const AlertsConnected = ({ className }) => (
  <StorageOverviewContext.Consumer>
    {props => <Alerts alertsResponse={props.alertsResponse} className={className} />}
  </StorageOverviewContext.Consumer>
);

AlertsConnected.propTypes = {
  ...Alerts.propTypes,
};

AlertsConnected.defaultProps = {
  ...Alerts.defaultProps,
};
