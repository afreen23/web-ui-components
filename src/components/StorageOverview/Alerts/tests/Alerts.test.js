import React from 'react';
import { shallow, render, mount } from 'enzyme';

import { Alerts, AlertsConnected } from '../Alerts';
import { default as AlertsFixtures } from '../fixtures/Alerts.fixture';
import { StorageOverviewContext } from '../../StorageOverviewContext';
import { selectDropdownItem } from '../../../../tests/enzyme';

import { ALL, WARNING, CRITICAL } from '../../../Dashboard/strings';

// eslint-disable-next-line react/prop-types
const testAlerts = ({ props }) => <Alerts {...props} />;

const getTypeDropdown = component => component.find('#alert-type');

const testAlertResults = (component, type) => {
  selectDropdownItem(getTypeDropdown(component), type);
  expect(component.state().type).toBe(type);
};

describe('<Alerts />', () => {
  AlertsFixtures.forEach(fixture => {
    it(`renders ${fixture.name} correctly`, () => {
      const component = shallow(testAlerts(fixture));
      expect(component).toMatchSnapshot();
    });
  });
  it('switches between alert dropdown options', () => {
    const component = mount(testAlerts(AlertsFixtures[0]));
    // default filter is ALL
    expect(component.state().type).toBe(ALL);
    testAlertResults(component, WARNING);
    testAlertResults(component, CRITICAL);
  });
  it('renders correctly with Provider', () => {
    const component = render(
      <StorageOverviewContext.Provider value={AlertsFixtures[0].props}>
        <AlertsConnected />
      </StorageOverviewContext.Provider>
    );
    expect(component).toMatchSnapshot();
  });
});
