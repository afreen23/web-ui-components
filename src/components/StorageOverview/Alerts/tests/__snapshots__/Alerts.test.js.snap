// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Alerts /> renders Alerts correctly 1`] = `
<DashboardCard
  className={null}
>
  <DashboardCardHeader
    className={null}
  >
    <Row
      bsClass="row"
      componentClass="div"
    >
      <Col
        bsClass="col"
        componentClass="div"
        lg={9}
        md={9}
        sm={9}
        xs={9}
      >
        <DashboardCardTitle
          className={null}
        >
          Alerts
        </DashboardCardTitle>
      </Col>
      <Col
        bsClass="col"
        componentClass="div"
        lg={3}
        md={3}
        sm={3}
        xs={3}
      >
        <DashboardCardActionsBody>
          <Dropdown
            choices={
              Array [
                "All",
                "Critical",
                "Warning",
              ]
            }
            disabled={false}
            id="alert-type"
            onChange={[Function]}
            value="All"
            withTooltips={false}
          />
        </DashboardCardActionsBody>
      </Col>
    </Row>
  </DashboardCardHeader>
  <DashboardCardBody
    LoadingComponent={[Function]}
    isLoading={false}
  >
    <AlertsBody>
      <AlertItem
        alert={
          Object {
            "annotations": Object {
              "description": "Disk inaccessible on host ip-10-0-154-62.ec2.internal (device xvda)",
              "message": "Disk is inaccessible",
              "storage_type": "ceph",
            },
            "labels": Object {
              "alertname": "CephOSDDiskUnavailable",
              "severity": "critical",
            },
          }
        }
        key="alert-0"
      />
      <AlertItem
        alert={
          Object {
            "annotations": Object {
              "description": "Data recovery has been active for over 2h. Contact Support",
              "message": "Data recovery is slow",
              "storage_type": "ceph",
            },
            "labels": Object {
              "alertname": "CephDataRecoveryTakingTooLong",
              "severity": "warning",
            },
          }
        }
        key="alert-1"
      />
    </AlertsBody>
  </DashboardCardBody>
</DashboardCard>
`;

exports[`<Alerts /> renders Data Recovery correctly 1`] = `
<DashboardCard
  className={null}
>
  <DashboardCardHeader
    className={null}
  >
    <Row
      bsClass="row"
      componentClass="div"
    >
      <Col
        bsClass="col"
        componentClass="div"
        lg={9}
        md={9}
        sm={9}
        xs={9}
      >
        <DashboardCardTitle
          className={null}
        >
          Alerts
        </DashboardCardTitle>
      </Col>
      <Col
        bsClass="col"
        componentClass="div"
        lg={3}
        md={3}
        sm={3}
        xs={3}
      >
        <DashboardCardActionsBody>
          <Dropdown
            choices={
              Array [
                "All",
                "Critical",
                "Warning",
              ]
            }
            disabled={false}
            id="alert-type"
            onChange={[Function]}
            value="All"
            withTooltips={false}
          />
        </DashboardCardActionsBody>
      </Col>
    </Row>
  </DashboardCardHeader>
  <DashboardCardBody
    LoadingComponent={[Function]}
    isLoading={false}
  >
    <AlertsBody>
      <AlertItem
        alert={
          Object {
            "annotations": Object {
              "description": "Data recovery has been active for over 2h. Contact Support",
              "message": "Data recovery is slow",
              "storage_type": "ceph",
            },
            "labels": Object {
              "alertname": "CephDataRecoveryTakingTooLong",
              "severity": "warning",
            },
          }
        }
        key="alert-0"
      />
    </AlertsBody>
  </DashboardCardBody>
</DashboardCard>
`;

exports[`<Alerts /> renders Disk Inaccessible correctly 1`] = `
<DashboardCard
  className={null}
>
  <DashboardCardHeader
    className={null}
  >
    <Row
      bsClass="row"
      componentClass="div"
    >
      <Col
        bsClass="col"
        componentClass="div"
        lg={9}
        md={9}
        sm={9}
        xs={9}
      >
        <DashboardCardTitle
          className={null}
        >
          Alerts
        </DashboardCardTitle>
      </Col>
      <Col
        bsClass="col"
        componentClass="div"
        lg={3}
        md={3}
        sm={3}
        xs={3}
      >
        <DashboardCardActionsBody>
          <Dropdown
            choices={
              Array [
                "All",
                "Critical",
                "Warning",
              ]
            }
            disabled={false}
            id="alert-type"
            onChange={[Function]}
            value="All"
            withTooltips={false}
          />
        </DashboardCardActionsBody>
      </Col>
    </Row>
  </DashboardCardHeader>
  <DashboardCardBody
    LoadingComponent={[Function]}
    isLoading={false}
  >
    <AlertsBody>
      <AlertItem
        alert={
          Object {
            "annotations": Object {
              "description": "Disk inaccessible on host ip-10-0-154-62.ec2.internal (device xvda)",
              "message": "Disk is inaccessible",
              "storage_type": "ceph",
            },
            "labels": Object {
              "alertname": "CephOSDDiskUnavailable",
              "severity": "critical",
            },
          }
        }
        key="alert-0"
      />
    </AlertsBody>
  </DashboardCardBody>
</DashboardCard>
`;

exports[`<Alerts /> renders Empty Alerts correctly 1`] = `""`;

exports[`<Alerts /> renders correctly with Provider 1`] = `
<div
  class="card-pf kubevirt-dashboard__card"
>
  <div
    class="card-pf-heading kubevirt-dashboard__card-heading"
  >
    <div
      class="row"
    >
      <div
        class="col-lg-9 col-md-9 col-sm-9 col-xs-9"
      >
        <h2
          class="card-pf-title kubevirt-dashboard__card-title"
        >
          Alerts
        </h2>
      </div>
      <div
        class="col-lg-3 col-md-3 col-sm-3 col-xs-3"
      >
        <div
          class="kubevirt-dashboard__actions-body"
        >
          <div
            class="btn-group btn-group-justified"
          >
            <div
              class="dropdown btn-group btn-group-default"
            >
              <button
                aria-expanded="false"
                aria-haspopup="true"
                class="kubevirt-dropdown dropdown-toggle btn btn-default"
                id="alert-type"
                role="button"
                type="button"
              >
                All 
                <span
                  class="caret"
                />
              </button>
              <ul
                aria-labelledby="alert-type"
                class="dropdown-menu"
                role="menu"
              >
                <li
                  class=""
                  role="presentation"
                >
                  <a
                    href="#"
                    role="menuitem"
                    tabindex="-1"
                  >
                    All
                  </a>
                </li>
                <li
                  class=""
                  role="presentation"
                >
                  <a
                    href="#"
                    role="menuitem"
                    tabindex="-1"
                  >
                    Critical
                  </a>
                </li>
                <li
                  class=""
                  role="presentation"
                >
                  <a
                    href="#"
                    role="menuitem"
                    tabindex="-1"
                  >
                    Warning
                  </a>
                </li>
              </ul>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="card-pf-body"
  >
    <div
      class="kubevirt-alert__alerts-body"
    >
      <div
        class="kubevirt-alert__item"
      >
        <span
          aria-hidden="true"
          class="fa fa-exclamation-circle kubevirt-alert__icon kubevirt-alert__icon--critical"
        />
        <div
          class="kubevirt-alert__item-message"
        >
          Disk inaccessible on host ip-10-0-154-62.ec2.internal (device xvda)
        </div>
      </div>
      <div
        class="kubevirt-alert__item"
      >
        <span
          aria-hidden="true"
          class="fa fa-exclamation-triangle kubevirt-alert__icon kubevirt-alert__icon--warning"
        />
        <div
          class="kubevirt-alert__item-message"
        >
          Data recovery has been active for over 2h. Contact Support
        </div>
      </div>
    </div>
  </div>
</div>
`;
